delta_shooter/game_mechanics.py

```diff
@@ -0,0 +1,314 @@
+
+GAME_SIZE = (600, 450)
+
+SPAWN_POINTS = [
+    (int(GAME_SIZE[0] * 0.1), GAME_SIZE[1] // 2),
+    (int(GAME_SIZE[0] * 0.9), GAME_SIZE[1] // 2),
+    (GAME_SIZE[0] // 2, int(GAME_SIZE[1] * 0.1)),
+    (GAME_SIZE[0] // 2, int(GAME_SIZE[1] * 0.9)),
+]
+
+SPAWN_ORIENTATIONS = [RIGHT, LEFT, DOWN, UP]
+
+
+HERE = Path(__file__).parent.resolve()
+
+
+def play_shooter(
+    your_choose_move: Callable[[np.ndarray], int],
+    opponent_choose_move: Callable[[np.ndarray], int],
+    game_speed_multiplier: float = 1,
+    render: bool = False,
+) -> float:
+    """Play a game where moves are chosen by `your_choose_move()` and `opponent_choose_move()`.
+
+    . You can render the game by setting `render=True`.
+
+    Args:
+        your_choose_move: function that chooses move (takes state as input)
+        opponent_choose_move: function that picks your opponent's next move
+        render: whether to render the game graphically
+        game_speed_multiplier: multiplies the speed of the game. High == fast
+                               (only has an effect when verbose=True)
+
+    Returns: total_return, which is the sum of return from the game
+    """
+    total_return = 0.0
+    game = ShooterEnv(
+        opponent_choose_move, render=render, game_speed_multiplier=game_speed_multiplier
+    )
+
+    state, _, done, _ = game.reset()
+    while not done:
+        action = your_choose_move(state)
+        state, reward, done, _ = game.step(action)
+        total_return += reward
+    return total_return
+
+
+def load_network(team_name: str, network_folder: Path = HERE) -> nn.Module:
+    net_path = network_folder / f"{team_name}_network.pt"
+    assert (
+        net_path.exists()
+    ), f"Network saved using TEAM_NAME='{team_name}' doesn't exist! ({net_path})"
+    model = torch.load(net_path)
+    model.eval()
+    return model
+
+
+def save_network(network: nn.Module, team_name: str) -> None:
+    assert isinstance(
+        network, nn.Module
+    ), f"train() function outputs an network type: {type(network)}"
+    assert "/" not in team_name, "Invalid TEAM_NAME. '/' are illegal in TEAM_NAME"
+    net_path = HERE / f"{team_name}_network.pt"
+    n_retries = 5
+    for attempt in range(n_retries):
+        try:
+            torch.save(network, net_path)
+            load_network(team_name)
+            return
+        except Exception:
+            if attempt == n_retries - 1:
+                raise
+
+
+def choose_move_randomly(state: np.ndarray) -> int:
+    return np.random.randint(4)
+
+
+class ShooterEnv(gym.Env):
+    def __init__(
+        self,
+        opponent_choose_move: Callable,
+        render: bool = False,
+        game_speed_multiplier: float = 1,
+    ):
+
+        self._render = render
+        self.opponent_choose_move = opponent_choose_move
+        self.game_speed_multiplier = game_speed_multiplier
+        if self._render:
+            self.init_graphics()
+        else:
+            self.screen = DummyScreen(GAME_SIZE)
```

Comment:
Have to also import `Union` from `typing`. This placates my type checker lower down, though.

```suggestion
            self.screen: Union[DummyScreen, pygame.Surface] = DummyScreen(GAME_SIZE)
```

Reply:
Interestingly my checker complains if i add this but doesn't complain if i don't

Reply:
:astonished:

Reply:
Okay, leave it as is :shrug: